<a href="https://colab.research.google.com/github/Spokai3/PhanTichDuLieuVaHocSau/blob/main/Lab8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

***1.Giới thiệu về thư viện NLTK:***

Cài đặt thư viện NLTK:

Vào cửa sổ command line, gõ lệnh pip install NLTK

Import thư viện NLTK và download công cụ NLTK

In [1]:
import nltk
nltk.download_shell()

NLTK Downloader
---------------------------------------------------------------------------
    d) Download   l) List    u) Update   c) Config   h) Help   q) Quit
---------------------------------------------------------------------------
Downloader> l

Packages:
  [ ] abc................. Australian Broadcasting Commission 2006
  [ ] alpino.............. Alpino Dutch Treebank
  [ ] averaged_perceptron_tagger Averaged Perceptron Tagger
  [ ] averaged_perceptron_tagger_ru Averaged Perceptron Tagger (Russian)
  [ ] basque_grammars..... Grammars for Basque
  [ ] biocreative_ppi..... BioCreAtIvE (Critical Assessment of Information
                           Extraction Systems in Biology)
  [ ] bllip_wsj_no_aux.... BLLIP Parser: WSJ Model
  [ ] book_grammars....... Grammars from NLTK Book
  [ ] brown............... Brown Corpus
  [ ] brown_tei........... Brown Corpus (TEI XML Version)
  [ ] cess_cat............ CESS-CAT Treebank
  [ ] cess_esp............ CESS-ESP Treebank
  [ ] chat80.....

tải gói gutenberg

In [2]:
nltk.download('gutenberg')

[nltk_data] Downloading package gutenberg to /root/nltk_data...
[nltk_data]   Package gutenberg is already up-to-date!


True

xem tên các tập tin có trong 'gutenberg' bằng lệnh fileids()

In [3]:
gb=nltk.corpus.gutenberg
print('Gutenberg files : ', gb.fileids())

Gutenberg files :  ['austen-emma.txt', 'austen-persuasion.txt', 'austen-sense.txt', 'bible-kjv.txt', 'blake-poems.txt', 'bryant-stories.txt', 'burgess-busterbrown.txt', 'carroll-alice.txt', 'chesterton-ball.txt', 'chesterton-brown.txt', 'chesterton-thursday.txt', 'edgeworth-parents.txt', 'melville-moby_dick.txt', 'milton-paradise.txt', 'shakespeare-caesar.txt', 'shakespeare-hamlet.txt', 'shakespeare-macbeth.txt', 'whitman-leaves.txt']


truy cập nội dung Macbeth của Shakespeare, sử dụng hàm words(), rồi gán cho một biến

In [4]:
macbeth = nltk.corpus.gutenberg.words('shakespeare-macbeth.txt')

dùng hàm len() để biết độ dài văn bản

In [5]:
len(macbeth)

23140

hiển thị 10 từ đầu tiên của tập tin

In [6]:
macbeth[:10]

['[',
 'The',
 'Tragedie',
 'of',
 'Macbeth',
 'by',
 'William',
 'Shakespeare',
 '1603',
 ']']

trích 5 câu đầu tiên của tập tin (1 câu được kẹp trong cặp ngoặc vuông), ta dùng hàm sent()

In [10]:
nltk.download('punkt')
macbeth_sents = nltk.corpus.gutenberg.sents('shakespeare-macbeth.txt')
macbeth_sents[:5]

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


[['[',
  'The',
  'Tragedie',
  'of',
  'Macbeth',
  'by',
  'William',
  'Shakespeare',
  '1603',
  ']'],
 ['Actus', 'Primus', '.'],
 ['Scoena', 'Prima', '.'],
 ['Thunder', 'and', 'Lightning', '.'],
 ['Enter', 'three', 'Witches', '.']]

***2.Tìm 1 từ với NLTK:***

Tìm từ 'Stage' xuất hiện trong văn bản text

In [12]:
text = nltk.Text(macbeth)
text.concordance('Stage')

Displaying 3 of 3 matches:
nts with Dishes and Seruice ouer the Stage . Then enter Macbeth Macb . If it we
with mans Act , Threatens his bloody Stage : byth ' Clock ' tis Day , And yet d
 struts and frets his houre vpon the Stage , And then is heard no more . It is 


Tìm từ xuất hiện trước và sau từ 'Stage'

In [13]:
text.common_contexts(['Stage'])

the_. bloody_: the_,


Tìm từ tương tự từ 'Stage'

In [14]:
text.similar('Stage')

day time face warre ayre king bleeding man reuolt serieant like
knowledge broyle shew head spring heeles hare thane skie


***3.Phân tích tần số của các từ***

muốn xem 10 từ thông dụng nhất trong văn bản xuất hiện bao nhiêu lần, dùng lệnh most_common()

In [15]:
fd = nltk.FreqDist(macbeth)
fd.most_common(10)

[(',', 1962),
 ('.', 1235),
 ("'", 637),
 ('the', 531),
 (':', 477),
 ('and', 376),
 ('I', 333),
 ('of', 315),
 ('to', 311),
 ('?', 241)]

Stopword là những từ thông dụng, thường ít có ý nghĩa trong quá trình phân tích văn bản và thường cần được loại bỏ.

In [16]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [17]:
sw = set(nltk.corpus.stopwords.words('english'))
print(len(sw))
list(sw)[:10]

179


['o',
 'against',
 'from',
 'few',
 'some',
 "mustn't",
 'don',
 "couldn't",
 "weren't",
 'you']

có 179 stopword trong từ vựng tiếng Anh. Ta sẽ loại bỏ các từ stopword trong biến macbeth

In [18]:
macbeth_filtered = [w for w in macbeth if w.lower() not in sw]
len(macbeth_filtered)

14946

In [19]:
fd = nltk.FreqDist(macbeth_filtered)
fd.most_common(10)

[(',', 1962),
 ('.', 1235),
 ("'", 637),
 (':', 477),
 ('?', 241),
 ('Macb', 137),
 ('haue', 117),
 ('-', 100),
 ('Enter', 80),
 ('thou', 63)]

loại bỏ các dấu câu theo lệnh

In [21]:
import string
punctuation = set(string.punctuation)
macbeth_filtered2 = [w.lower() for w in macbeth if w.lower() not in sw and w.lower() not in punctuation]
fd = nltk.FreqDist(macbeth_filtered2)
fd.most_common(10)

[('macb', 137),
 ('haue', 122),
 ('thou', 90),
 ('enter', 81),
 ('shall', 68),
 ('macbeth', 62),
 ('vpon', 62),
 ('thee', 61),
 ('macd', 58),
 ('vs', 57)]

***4.Lựa chọn các từ trong văn bản***

Rút trích các từ có độ dài nhất, ví dụ các từ có độ dài lớn hơn 12 ký tự, dùng lệnh:

In [22]:
long_words = [w for w in macbeth if len(w) >12]
sorted(long_words)

['Assassination',
 'Chamberlaines',
 'Distinguishes',
 'Gallowgrosses',
 'Metaphysicall',
 'Northumberland',
 'Voluptuousnesse',
 'commendations',
 'multitudinous',
 'supernaturall',
 'vnaccompanied']

Rút trích các từ có chứa chuỗi 'ious'


In [23]:
ious_words = [w for w in macbeth if 'ious' in w]
ious_words = set(ious_words)
sorted(ious_words)

['Auaricious',
 'Gracious',
 'Industrious',
 'Iudicious',
 'Luxurious',
 'Malicious',
 'Obliuious',
 'Pious',
 'Rebellious',
 'compunctious',
 'furious',
 'gracious',
 'pernicious',
 'pernitious',
 'pious',
 'precious',
 'rebellious',
 'sacrilegious',
 'serious',
 'spacious',
 'tedious']

***5.Bigrams và collocations***

Lọc các bigram sau khi đã loại các stopword và các dấu câu, dùng lệnh sau:

In [24]:
bgrms = nltk.FreqDist(nltk.bigrams(macbeth_filtered2))
bgrms.most_common(15)

[(('enter', 'macbeth'), 16),
 (('exeunt', 'scena'), 15),
 (('thane', 'cawdor'), 13),
 (('knock', 'knock'), 10),
 (('st', 'thou'), 9),
 (('thou', 'art'), 9),
 (('lord', 'macb'), 9),
 (('haue', 'done'), 8),
 (('macb', 'haue'), 8),
 (('good', 'lord'), 8),
 (('let', 'vs'), 7),
 (('enter', 'lady'), 7),
 (('wee', 'l'), 7),
 (('would', 'st'), 6),
 (('macbeth', 'macb'), 6)]

Ngoài bigram ra, còn có trigram, sự kết hợp của 3 từ, ta dùng lệnh trigrams()

In [25]:
tgrms = nltk.FreqDist(nltk.trigrams (macbeth_filtered2))
tgrms.most_common(10)

[(('knock', 'knock', 'knock'), 6),
 (('enter', 'macbeth', 'macb'), 5),
 (('enter', 'three', 'witches'), 4),
 (('exeunt', 'scena', 'secunda'), 4),
 (('good', 'lord', 'macb'), 4),
 (('three', 'witches', '1'), 3),
 (('exeunt', 'scena', 'tertia'), 3),
 (('thunder', 'enter', 'three'), 3),
 (('exeunt', 'scena', 'quarta'), 3),
 (('scena', 'prima', 'enter'), 3)]

***6.Sử dụng văn bản trên mạng***

import thư viện và mở url để đọc file

In [26]:
from urllib import request
url = 'http://www.gutenberg.org/files/2554/2554-0.txt'
response = request.urlopen(url)
raw = response.read().decode('utf8')
raw[:75]

'\ufeffThe Project Gutenberg eBook of Crime and Punishment, by Fyodor Dostoevsky\r'

thay bằng các lệnh sau:

In [29]:
from urllib import request
url = 'http://www.gutenberg.org/files/2554/2554-0.txt'
response = request.urlopen(url)
raw = response.read().decode('utf-8-sig')
raw[:75]

'The Project Gutenberg eBook of Crime and Punishment, by Fyodor Dostoevsky\r\n'

In [30]:
tokens = nltk.word_tokenize(raw)
webtext = nltk.Text(tokens)
webtext[:12]

['The',
 'Project',
 'Gutenberg',
 'eBook',
 'of',
 'Crime',
 'and',
 'Punishment',
 ',',
 'by',
 'Fyodor',
 'Dostoevsky']

***7.Rút trích văn bản từ trang html***

cách rút ttich1 văn bản từ trang html

In [31]:
url = "http://news.bbc.co.uk/2/hi/health/2284783.stm"
html = request.urlopen(url).read().decode('utf8')
html[:120]

'<!doctype html public "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">\r\n<html>\r\n<hea'

dùng thư viện bs4(BeautifulSoup) cung cấp cho bạn các trình phân tích cú pháp phù hợp có thể nhận dạng HTML và trích xuất văn bản

In [32]:
from bs4 import BeautifulSoup
raw = BeautifulSoup(html, 'lxml').get_text()
tokens = nltk.word_tokenize(raw)
text = nltk.Text(tokens)

***8.Phân tích cảm xúc người dùng***

download các movie review dùng lệnh sau:

In [33]:
nltk.download('movie_reviews')

[nltk_data] Downloading package movie_reviews to /root/nltk_data...
[nltk_data]   Unzipping corpora/movie_reviews.zip.


True

In [34]:
import random 
reviews = nltk.corpus.movie_reviews
documents = [(list(reviews.words(fileid)), category)
for category in reviews.categories()
for fileid in reviews.fileids(category)]
random.shuffle(documents)

Xem nội dung review đầu tiên (dòng 0, cột 0)

In [35]:
first_review = ' '.join(documents[0][0])
print(first_review)

the characters in " palmetto " collectively sweat enough to fill lake erie , and it ' s just the right way to capture the film ' s ripe atmosphere of sensuality , tension and even farce . this humid florida - set thriller nurtures a modestly absurd edginess that often winks at itself , which is very much a good thing since the actors and director are in on the joke . as far as set - up is concerned , the movie is nothing we haven ' t seen before , but be patient -- " palmetto " spins some pretty wicked story twists that we really don ' t see coming . it ' s not a classic by any means , but it sure is quite an engaging surprise . harry barber ( woody harrelson ) is fresh out of prison , where he ' s served a couple of years for a crime he didn ' t commit . even though he harbors animosity towards the obtuse local officials , harry ' s first order of business is to return to his home town of palmetto to be with his sculptor girlfriend nina ( gina gershon ) . while searching for work , he

Xem kết quả review đầu tiên (dòng 0, cột 1)

In [36]:
documents[0][1]

'pos'

Ta cần tạo bảng phân phối tần số các từ trong copus, bảng này cần chuyển sang dạng list, ta dùng hàm list()

In [38]:
all_words = nltk.FreqDist(w.lower() for w in reviews.words())
word_features = list(all_words)

Bước tiếp theo là xác dịnh một hàm để tính toán các đặc trưng, tức là những từ đủ quan trọng để thiết lập ý kiến của một review

In [43]:
def document_features(document, word_features):
    document_words = set(document)
    features = {}
    for word in word_features:
        features['{}', format(word)] = (word in document_words)
    return features

Khi định nghĩa hàm document_features(), bạn tạo 1 tập các documents

In [44]:
featuresets = [(document_features(d,word_features), c) for (d,c) in documents]
len(featuresets)

2000

Tạo tập train và test: 1500 dòng đầu dùng cho tập train và 500 dòng cho tập test để đánh giá độ chính xác của mô hình

In [45]:
train_set, test_set = featuresets[1500:], featuresets[:500]
classifier = nltk.NaiveBayesClassifier.train(train_set)

Dùng thuật toán Naive Bayes để phân loại, dùng thư viện NLTK. Sau đó tính toán độ chính xác của thuật toán

In [46]:
train_set, test_set = featuresets[1500:],featuresets[:500]
classifier= nltk.NaiveBayesClassifier.train(train_set)
print(nltk.classify.accuracy(classifier, test_set))

0.744


Chúng ta đã hoàn tất việc phân tích, dưới đây các từ với trong số lớn nhất của các review được đánh giá là positive và negative

In [47]:
classifier.show_most_informative_features(10)

Most Informative Features
   ('{}', 'outstanding') = True              pos : neg    =     12.1 : 1.0
   ('{}', 'perspective') = True              pos : neg    =     11.4 : 1.0
    ('{}', 'delightful') = True              pos : neg    =      9.2 : 1.0
   ('{}', 'potentially') = True              neg : pos    =      8.2 : 1.0
 ('{}', 'uninteresting') = True              neg : pos    =      8.2 : 1.0
        ('{}', 'smooth') = True              pos : neg    =      7.7 : 1.0
        ('{}', 'enters') = True              pos : neg    =      7.7 : 1.0
      ('{}', 'identify') = True              pos : neg    =      7.7 : 1.0
     ('{}', 'energetic') = True              pos : neg    =      7.7 : 1.0
        ('{}', 'wasted') = True              neg : pos    =      7.1 : 1.0
